[RFT] uboot-mediatek: update to v2025.01

DragonBluep opened the pull request:
Update u-boot for Mediatek to v2025.01 and backport some mtk_eth fixes. cc @dangowrt @Ansuel

DragonBluep committed:
uboot-mediatek: update to v2025.01

Remove upstreamed patches:
010-menu-fix-the-logic-checking-whether-ESC-key-is-press.patch [1]
011-menu-add-support-to-check-if-menu-needs-to-be-reprin.patch [2]
012-bootmenu-add-reprint-check.patch [3]

Remove outdated patches:
455-arm-provide-noncached_set_region-prototype-to-fix-build.patch

Some patches have been manually rebased to match the upstream
changes. This patch also fixes the dtc warning for reserved-memory
dts node. If #address-cells and #size-cells are not same as the
root node definitions, the dtc will complain about it.

All defconfigs are refreshed by `make "$board"_defconfig` and
`make savedefconfig`.

[1] https://github.com/u-boot/u-boot/commit/ddac69885efaffc68a71c1a159dc2b3ff4b9fda3
[2] https://github.com/u-boot/u-boot/commit/ccdd7948e22f21d2add8f51c4918a2c576dc5e91
[3] https://github.com/u-boot/u-boot/commit/599652cff13ad31d23b1b8bf8905533e447435d5

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
uboot-mediatek: backport mtk_eth fixes from u-boot next

This patch series will provide better support for Mediatek
ethernet and add support for Airoha AN8855.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

mrkiko commented:
This is great!! Going to test it soon in GDSP.

DragonBluep committed:
uboot-mediatek: sync mtk-openwrt open source u-boot patches

Synchronize the latest MTK u-boot patches[1]. Some patches have
been amended since last synchronization.

Changes:
* Minor NMBM layer fixes and improvements.
* A new bootmenu shortkey implementation.
* New SPI flash support for en25qx128.

[1] https://github.com/mtk-openwrt/u-boot/tree/mtksoc-20230719

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

Ansuel commented:
thanks a lot will test this soon on openwrt one

csharper2005 commented:
> backport some mtk_eth fixes.

Is there any chances add the following series to your PR?
https://patchwork-proxy.ozlabs.org/project/uboot/list/?series=439654

sc16me commented:
It would be great to add
https://patchwork-proxy.ozlabs.org/project/uboot/patch/e541ccc8ef00bf33ed0fd897f4f207d44d0cb5ad.1736498083.git.weijie.gao@mediatek.com/

DragonBluep committed:
uboot-mediatek: backport more mtk platform pending fixes

* Cumulative fixes and updates for MediaTek platform. [1]
* Update mt7981 pinctrl driver based on upstream kernel. [2]

[1] https://lore.kernel.org/u-boot/cover.1737104723.git.weijie.gao@mediatek.com/
[2] https://lore.kernel.org/u-boot/20250124033902.187796-1-weijie.gao@mediatek.com/

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
> > backport some mtk_eth fixes.
> 
> Is there any chances add the following series to your PR? https://patchwork-proxy.ozlabs.org/project/uboot/list/?series=439654

Pulled. The new AN8855 driver will be automatically selected for MT7981.

csharper2005 commented:
> Pulled. The new AN8855 driver will be automatically selected for MT7981.

Thanks a lot! It works. Xiaomi AX3000t with Airoha AN8855.

P.S. I had to change `mediatek,switch = "mt7531";` to `mediatek,switch = "an8855";` here - https://github.com/openwrt/openwrt/blob/main/package/boot/uboot-mediatek/patches/440-add-xiaomi_mi-router-ax3000t.patch#L182 I wonder if we can combine both switches in one dts or we need separate `mtk` and `airoha` versions of u-boot.

DragonBluep commented:
> > Pulled. The new AN8855 driver will be automatically selected for MT7981.
> 
> Thanks a lot! It works. Xiaomi AX3000t with Airoha AN8855.
> 
> P.S. I had to change `mediatek,switch = "mt7531";` to `mediatek,switch = "an8855";` here - https://github.com/openwrt/openwrt/blob/main/package/boot/uboot-mediatek/patches/440-add-xiaomi_mi-router-ax3000t.patch#L182 I wonder if we can combine both switches in one dts or we need separate `mtk` and `airoha` versions of u-boot.

We can use `mediatek,switch = "auto";` to support both of them.

csharper2005 commented:
> We can use `mediatek,switch = "auto";` to support both of them.

This is true.
```
F0: 102B 0000
FA: 1040 0000
FA: 1040 0000 [0200]
F9: 0000 0000
V0: 0000 0000 [0001]
00: 0000 0000
BP: 2400 0041 [0000]
G0: 1190 0000
EC: 0000 0000 [1000]
T0: 0000 0236 [010F]
Jump to BL

NOTICE:  BL2: v2.10.0   (release):OpenWrt v2024.01.17~bacca82a-3 (mt7981-spim-nand-ddr3)
NOTICE:  BL2: Built : 13:25:03, Jan 12 2025
NOTICE:  WDT: [40000000] Software reset (reboot)
NOTICE:  EMI: Using DDR3 settings
NOTICE:  EMI: Detected DRAM size: 256MB
NOTICE:  EMI: complex R/W mem test passed
NOTICE:  CPU: MT7981 (1300MHz)
NOTICE:  SPI_NAND parses attributes from parameter page.
NOTICE:  SPI_NAND Detected ID 0xef
NOTICE:  Page size 2048, Block size 131072, size 134217728
NOTICE:  BL2: Booting BL31
NOTICE:  BL31: v2.10.0  (release):OpenWrt v2024.01.17~bacca82a-3 (mt7981-spim-nand-ddr3)
NOTICE:  BL31: Built : 13:25:03, Jan 12 2025


U-Boot 2025.01-OpenWrt-r28539+3-e1ff90098e (Jan 24 2025 - 20:06:30 +0000)

CPU:   MediaTek MT7981
Model: Xiaomi Router AX3000T
DRAM:  256 MiB
Core:  39 devices, 18 uclasses, devicetree: embed
Loading Environment from UBI... spi-nand: spi_nand spi_nand@0: Winbond SPI NAND was found.
spi-nand: spi_nand spi_nand@0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
Read 126976 bytes from volume ubootenv to 000000004f7c1c80
Read 126976 bytes from volume ubootenv2 to 000000004f7e0cc0
OK
In:    serial@11002000
Out:   serial@11002000
Err:   serial@11002000
reset button found
Loading Environment from UBI... UBI partition 'ubi' already selected
Read 126976 bytes from volume ubootenv to 000000004f7c1c80
Read 126976 bytes from volume ubootenv2 to 000000004f7e0cc0
OK
Net:   Airoha AN8855

Warning: ethernet@15100000 (eth0) using random MAC address - 1a:88:ff:94:cf:5e
eth0: ethernet@15100000

        ( ( ( OpenWrt ) ) )       U-Boot 2025.01-OpenWrt-r28539+2-e1ff90098e (Ja
      1. Run default boot command.
      2. Boot system via TFTP.
      3. Boot production system from NAND.
      4. Boot recovery system from NAND.
      5. Load production system via TFTP then write to NAND.
      6. Load recovery system via TFTP then write to NAND.
      7. Load BL31+U-Boot FIP via TFTP then write to NAND.
      8. Load BL2 preloader via TFTP then write to NAND.
      9. Reboot.
      a. Reset all settings to factory defaults.
      0. Exit


  Press UP/DOWN to move, ENTER to select, ESC to quit
Using ethernet@15100000 device
TFTP from server 192.168.1.254; our IP address is 192.168.1.1
Filename 'openwrt-mediatek-filogic-xiaomi_mi-router-ax3000t-ubootmod-initramfs-recovery.itb'.
Load address: 0x46000000
Loading: #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         ##################################################
         1.2 MiB/s
done
Bytes transferred = 7405568 (710000 hex)
## Loading kernel from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'kernel-1' kernel subimage
     Description:  ARM64 OpenWrt Linux-6.6.73
     Type:         Kernel Image
     Compression:  lzma compressed
     Data Start:   0x460000e8
     Data Size:    4124844 Bytes = 3.9 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: 0x48000000
     Entry Point:  0x48000000
     Hash algo:    crc32
     Hash value:   11ac8bff
     Hash algo:    sha1
     Hash value:   ae926497fd98aa52b48e95edfa50913e5d858c66
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Loading ramdisk from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'initrd-1' ramdisk subimage
     Description:  ARM64 OpenWrt xiaomi_mi-router-ax3000t-ubootmod initrd
     Type:         RAMDisk Image
     Compression:  uncompressed
     Data Start:   0x463ef2e0
     Data Size:    3197352 Bytes = 3 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: unavailable
     Entry Point:  unavailable
     Hash algo:    crc32
     Hash value:   78a3150f
     Hash algo:    sha1
     Hash value:   51457e68ef783dbc604d902468cb96185e39f312
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Loading fdt from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'fdt-1' fdt subimage
     Description:  ARM64 OpenWrt xiaomi_mi-router-ax3000t-ubootmod device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x466fbda8
     Data Size:    28061 Bytes = 27.4 KiB
     Architecture: AArch64
     Hash algo:    crc32
     Hash value:   7df3fae3
     Hash algo:    sha1
     Hash value:   d2b14f1a27c4e0eca3b254a703386b3ea9090769
   Verifying Hash Integrity ... crc32+ sha1+ OK
   Booting using the fdt blob at 0x466fbda8
Working FDT set to 466fbda8
   Uncompressing Kernel Image to 48000000
   Loading Ramdisk to 4e4c3000, end 4e7cf9a8 ... OK
   Loading Device Tree to 000000004e4b9000, end 000000004e4c2d9c ... OK
Working FDT set to 4e4b9000
Add 'ramoops@42ff0000' node failed: FDT_ERR_EXISTS

Starting kernel ...

<6>[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
<5>[    0.000000] Linux version 6.6.73 (builder@buildhost) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 13.3.0 r28304-6dacba30a7) 13.3.0, GNU ld (GNU Binutils) 2.42) #0 SMP Wed Jan 22 19:52:54 2025
```

DragonBluep committed:
uboot-mediatek: enable auto switch chip detect for Xiaomi AX3000T

The Xiaomi AX3000T has two hardware revisions. One uses MT7531
switch, and the other uses AN8855 switch. Set "mediatek,switch"
property to "auto" to be compatible with different switches.

Tested-by: Mikhail Zhilkin <csharper2005@gmail.com>
Signed-off-by: Shiji Yang <yangshiji66@qq.com>

ServError committed:
uboot-mediatek: port patches for FORESEE NAND support

Backport FORESEE NAND chip support from upstream Linux. The newly
introduced FORESEE F35SQA001G was found on the Xiaomi AX3000T.

Signed-off-by: Erik Servili <serverror@serverror.com>
Signed-off-by: Shiji Yang <yangshiji66@qq.com>

mrkiko commented:
I am testing this right now. I would like to ask if it could be possible to all headers to all patches, to allow for correct
"git am"
application. This makes it easier to test things out if you need to keep some - albeit small - out of tree patches.

mrkiko commented:
OK - I can confirm this works on the GDSP device.

Tested-by: Enrico Mioso <mrkiko.rs@gmail.com>

DragonBluep committed:
uboot-mediatek: move default env files to a dedicated directory

Putting all files in the root folder is too messy.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

csharper2005 commented:
```
[Xiaomi AX3000t: Winbond, Airoha AN8855]
[Routerich AX3000]
Tested-by: Mikhail Zhilkin <csharper2005@gmail.com>
```

ServError commented:
@DragonBluep Could you review the PR against your fork https://github.com/DragonBluep/openwrt/pull/1 and include the changes with this PR to round out the Xiaomi AX3000T support across hardware variants?

DragonBluep commented:
> @DragonBluep Could you review the PR against your fork [DragonBluep#1](https://github.com/DragonBluep/openwrt/pull/1) and include the changes with this PR to round out the Xiaomi AX3000T support across hardware variants?

@ServError Thanks! I merged it into this patch series.

All supported boards are only build tested. MT7981 and MT762x mips SoCs should work as well as before.

VoidNullNil commented:
I merged this PR into my build over 24.10, and Foresee worked flawlessly. Hoping this PR gets merged soon!

dangowrt commented:
Tested on MT7621, MT7622, MT7986 and MT7988, all works fine. Merging this now.

danpawlik commented:
almost all is fine on BPi-R4 - it just ignore mem limitation that is helpful to enable WED.
I mean:
```
# previous u-boot 
$ fw_printenv | grep bootargs
console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait 

# uboot 2025.01
$ fw_printenv | grep bootargs
console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait ubi.block=0,fit

# enabling 2GB limitation to enable WED
fw_setenv bootargs 'console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait ubi.block=0,fit mem=2048'
```
And the RAM limitation does not work. Is it any other way how to enable it?

dangowrt commented:
> Is it any other way how to enable it?

Can you check in `dmesg` is the kernel command line contains the `mem=2048` argument?
Also check that `bootargs` U-Boot environment variable is persistent and present while booting the kernel.

hackpascal commented:
P.S. These patches have been tested on all supported SoCs, including mt7623, mt7629 and mt7981, before submitting to mainline.

danpawlik commented:
@dangowrt hey, thanks for reply.
It is present in bootargs and in dmesg:

```
root@BPI-R4:~# fw_printenv | grep bootargs
bootargs=console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 mem=2048M rootwait ubi.block=0,fit

root@BPI-R4:~# dmesg | grep 2048
[    0.000000] Kernel command line: console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait ubi.block=0,fit mem=2048
[    0.027708] tcp_listen_portaddr_hash hash table entries: 2048 (order: 3, 32768 bytes, linear)
[    0.028700] UDP hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    0.028751] UDP-Lite hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    1.154983] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    2.130051] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    2.136829] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[   13.031410] UBIFS (ubi0:6): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
```

So issue maybe is somewhere else, but... why kernel command like shows just mem=2048? Odd.

dmsza commented:
Could this PR be backported to the 24.10 branch for inclusion in the upcoming 24.10.1 stable release?

DragonBluep commented:
> @dangowrt hey, thanks for reply. It is present in bootargs and in dmesg:
> 
> ```
> root@BPI-R4:~# fw_printenv | grep bootargs
> bootargs=console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 mem=2048M rootwait ubi.block=0,fit
> 
> root@BPI-R4:~# dmesg | grep 2048
> [    0.000000] Kernel command line: console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait ubi.block=0,fit mem=2048
> [    0.027708] tcp_listen_portaddr_hash hash table entries: 2048 (order: 3, 32768 bytes, linear)
> [    0.028700] UDP hash table entries: 2048 (order: 4, 65536 bytes, linear)
> [    0.028751] UDP-Lite hash table entries: 2048 (order: 4, 65536 bytes, linear)
> [    1.154983] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
> [    2.130051] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
> [    2.136829] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
> [   13.031410] UBIFS (ubi0:6): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
> ```
> 
> So issue maybe is somewhere else, but... why kernel command like shows just mem=2048? Odd.

It seems that your u-boot env config file is incorrect. A full reset should solve the issue for you.

danpawlik commented:
@DragonBluep you mean the option via UART: `Reset all settings to factory defaults` ?
I have done it already, then I just add `mem=2048M`, that's why I came here.

danpawlik commented:
tl;dr to set the limit on BPi-R4, I need to be on eMMC or SD card, otherwise it will not work.
